--- reader-videoguard2.c.orig	2022-06-21 21:43:48.395083296 +0200
+++ reader-videoguard2.c	2022-06-21 21:46:13.637123810 +0200
@@ -1187,7 +1187,6 @@
 static int32_t videoguard2_do_ecm(struct s_reader *reader, const ECM_REQUEST *er, struct s_ecm_answer *ea)
 {
 	uint8_t cta_res[CTA_RES_LEN];
-	static const char valid_ecm[] = { 0x00, 0x00, 0x01 };
 	uint8_t ins40[5] = { 0xD1, 0x40, 0x60, 0x80, 0xFF };
 	static const uint8_t ins54[5] = { 0xD3, 0x54, 0x00, 0x00, 0x00};
 	int32_t posECMpart2 = er->ecm[6] + 7;
@@ -1200,11 +1199,15 @@
 	int32_t chk;
 	chk = checksum_ok(EcmIrdHeader);
 
-	if((memcmp(&(er->ecm[3]), valid_ecm, sizeof(valid_ecm)) != 0) || (chk == 0))
+	if((er->ecm[5] != 1) || chk == 0)
 	{
 		rdr_log(reader, "Not a valid ecm");
 		return E_CORRUPT;
 	}
+	if(4 == er->ecm[4] - er->ecm[2]) //icam mode
+	{
+		lenECMpart2  -= 4;
+	}
 	memcpy(tbuff + 1, er->ecm + posECMpart2 + 1, lenECMpart2 - 1);
 
 	int32_t new_len = lenECMpart2;
